projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/util/LLVMControlFlowGraph.java

```diff
@@ -0,0 +1,335 @@
+         * the outer loop
+         */
+        private void calculateSuccessors() {
+            successors = new HashSet<>();
+
+            for (CFGBlock s : loopHeader.sucs) {
+                if (!isInLoop(s)) {
+                    successors.add(s);
+                }
+            }
+
+            for (CFGBlock b : body) {
+                // for each inner loop, add all successors which are not in the outer loop
+                // to the successors of the outer loop
+                if (b.isLoopHeader) {
+                    for (CFGLoop l : innerLoops) {
+                        if (l.getHeader().equals(b)) {
+                            for (CFGBlock ib : l.getSuccessors()) {
+                                if (!isInLoop(ib)) {
+                                    successors.add(ib);
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    for (CFGBlock s : b.sucs) {
+                        if (!isInLoop(s)) {
+                            successors.add(s);
+                        }
+                    }
+                }
+            }
+
+        }
+
+        public Integer[] getSuccessorIDs() {
+            if (successors == null) {
+                calculateSuccessors();
+            }
+
+            Integer[] sucIDs = new Integer[successors.size()];
+            int i = 0;
+            for (CFGBlock b : successors) {
+                sucIDs[i++] = b.id;
+            }
+
+            return sucIDs;
+        }
+
+        public boolean isInLoop(CFGBlock block) {
+            return block == loopHeader || body.contains(block);
+        }
+
+        public CFGBlock getHeader() {
+            return loopHeader;
+        }
+
+        @Override
+        public String toString() {
+            return "Loop: " + this.id + " - Header: " + this.loopHeader.toString();
+        }
+
+    }
+
+    public LLVMControlFlowGraph(InstructionBlock[] blocks) {
+        this.blocks = new CFGBlock[blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            this.blocks[i] = new CFGBlock(blocks[i]);
+        }
+    }
+
+    private void resolveEdges() {
+        for (CFGBlock block : blocks) {
+            TerminatingInstruction term = block.instructionBlock.getTerminatingInstruction();
+            // set successors and predecessors
+            for (int i = 0; i < term.getSuccessorCount(); i++) {
+                int sucId = term.getSuccessor(i).getBlockIndex();
+                block.sucs.add(this.blocks[sucId]);
+                this.blocks[sucId].preds.add(block);
+            }
+        }
+
+    }
+
+    public List<CFGLoop> getCFGLoops() {
+        List<CFGLoop> loops = new ArrayList<>();
+
+        for (CFGLoop l : cfgLoops) {
+            if (l != null)
+                loops.add(l);
+        }
+        loops.removeIf(l -> l == null);
```

Comment:
This statement is unnecessary since you already make sure not to add any `null` elements to the list.